### PR TITLE
migrate_event: Add case for migration event

### PR DESCRIPTION
**Pull request by cliping**
Case ID: RHEL-198888
Test scenarios:
Check events when cancelling migration in PerformPhase and after
src vm is paused.

Signed-off-by: lcheng <lcheng@redhat.com>

### CODE REVIEW

**Comment by cliping** [comment]
Test result:

#avocado run --vt-type libvirt --vt-machine-type q35 virsh.migrate_options_shared.positive_test.check_event.cancel_migration.without_postcopy --vt-connect-uri qemu:///system
WARNING:root:No python imaging library installed. Screendump and Windows guest BSOD detection are disabled. In order to enable it, please install python-imaging or the equivalent for your distro.
WARNING:root:No python imaging library installed. PPM image conversion to JPEG disabled. In order to enable it, please install python-imaging or the equivalent for your distro.
JOB ID     : f076221f0e12058c824c5649f5a30d7a2955f061
JOB LOG    : /root/avocado/job-results/job-2021-07-21T07.56-f076221/job.log
 (1/1) type_specific.io-github-autotest-libvirt.virsh.migrate_options_shared.positive_test.check_event.cancel_migration.without_postcopy: PASS (204.04 s)
RESULTS    : PASS 1 | ERROR 0 | FAIL 0 | SKIP 0 | WARN 0 | INTERRUPT 0 | CANCEL 0
JOB TIME   : 205.08 s

**Review by fangge1212 — APPROVED** [review]

File: libvirt/tests/src/migration/migrate_options_shared.py [hunk 1]
```diff
@@ -1185,9 +1198,10 @@ def cancel_bg_migration():
         cleanup_libvirtd_log(log_file)
 
         # Setup libvirt related conf files
-        update_qemu_conf_on_local_and_remote()
-        update_libvirtd_conf_on_local_and_remote()
-        update_log_conf_on_local_and_remote()
+        if not cancel_migration:
```
Comment (Yingshun) [inline libvirt/tests/src/migration/migrate_options_shared.py:1201]: update_xxx_conf_xxx() functions should do nothing if there is no configuration setting related variants. So is this change necessary?

File: libvirt/tests/cfg/migration/migrate_options_shared.cfg [hunk 2]
```diff
@@ -317,6 +317,20 @@
...
+                        - cancel_migration:
+                            check_event_output = "yes"
+                            virsh_migrate_options = "--live --p2p --verbose"
+                            virsh_migrate_extra = "--timeout 5 --timeout-suspend --bandwidth 5"
+                            grep_str_local_log = ""
+                            grep_str_remote_log = ""
+                            check_log_interval = "no"
+                            low_speed = "5"
+                            asynch_migrate = "yes"
+                            status_error = 'yes'
```
[the remark this answers is not in the document]
Comment (cliping) [reply libvirt/tests/cfg/migration/migrate_options_shared.cfg]: Update.

File: libvirt/tests/src/migration/migrate_options_shared.py [hunk 3]
```diff
@@ -984,6 +986,16 @@ def cancel_bg_migration():
         logging.debug("status:[%d], stdout:[%s], stderr:[%s]",
                       p.returncode, stdout, stderr)
 
+    def cancel_migration_performphase():
+        """
+        Cancel migration process in performphase
+        """
+        time.sleep(15)
```
[the remark this answers is not in the document]
Comment (cliping) [reply libvirt/tests/src/migration/migrate_options_shared.py]: Update.

**Comment by cliping** [comment]
Test result:

#avocado run --vt-type libvirt --vt-machine-type q35 virsh.migrate_options_shared.negative_test.p2p_migration.cancel_migration.without_postcopy --vt-connect-uri qemu:///system
WARNING:root:No python imaging library installed. Screendump and Windows guest BSOD detection are disabled. In order to enable it, please install python-imaging or the equivalent for your distro.
WARNING:root:No python imaging library installed. PPM image conversion to JPEG disabled. In order to enable it, please install python-imaging or the equivalent for your distro.
JOB ID     : 8aa6f98e3f74068c5c370487369c8be560136eec
JOB LOG    : /root/avocado/job-results/job-2021-07-23T06.38-8aa6f98/job.log
 (1/1) type_specific.io-github-autotest-libvirt.virsh.migrate_options_shared.negative_test.p2p_migration.cancel_migration.without_postcopy: PASS (170.51 s)
RESULTS    : PASS 1 | ERROR 0 | FAIL 0 | SKIP 0 | WARN 0 | INTERRUPT 0 | CANCEL 0
JOB TIME   : 171.02 s

File: libvirt/tests/src/migration/migrate_options_shared.py [hunk 4]
```diff
@@ -984,6 +988,30 @@ def cancel_bg_migration():
...
+        """
+        Check source event in migration processes
+        """
+        def check_event():
+            """Check event"""
+            output = virsh_session.get_stripped_output()
+            logging.debug("Actual output is %s", output)
+            if not re.findall("Suspended Migrated", output):
+                return False
+            return True
```
Comment (Yingshun) [inline libvirt/tests/src/migration/migrate_options_shared.py:1001]: I still think it can be covered by multi_funcs in do_migration() and the scope of this function is too narrow, it'll only research "Suspended Migrated" event... @dzhengfy what's your opinion?

File: libvirt/tests/cfg/migration/migrate_options_shared.cfg [hunk 5]
```diff
@@ -613,6 +613,15 @@
...
                                             err_msg = "error: internal error: qemu unexpectedly closed the monitor"
+                        - cancel_migration:
+                            check_event_output = "yes"
+                            virsh_migrate_extra = "--timeout 5 --timeout-suspend --bandwidth 5"
+                            check_log_interval = "no"
+                            low_speed = "5"
+                            asynch_migrate = "yes"
+                            actions_during_migration = "check_src_event,cancel_migration_in_performphase"
+                            expectedEventSrc = ["event 'migration-iteration'", "event 'lifecycle' for domain.*: Suspended Migrated", "event 'lifecycle' for domain .*: Resumed Migrated"]
+                            expectedEventTarget = ["event 'lifecycle' for domain.*Started Migrated", "event 'lifecycle' for domain.*: Stopped Failed"]
```
Comment (dzhengfy) [inline libvirt/tests/cfg/migration/migrate_options_shared.cfg:624]: Could you help explain where expectedEventSrc and expectedEventTarget are used? I did not find the reference.

File: libvirt/tests/src/migration/migrate_options_shared.py (hunk 4 above)
Comment (dzhengfy) [reply libvirt/tests/src/migration/migrate_options_shared.py:1001]: yes. If your purpose is to wait for the event 'Suspended Migrated' which means the vm is paused, please refer to doc in do_migration(xxx, multi_funcs, xxx) in migration.py. Or your purpose is only to do actions after vm is paused,  it is also supported by using "before_pause": "no".

File: libvirt/tests/src/migration/migrate_options_shared.py [hunk 6]
```diff
@@ -984,6 +988,30 @@ def cancel_bg_migration():
...
+        check_ret = utils_misc.wait_for(check_event, timeout=20)
+        if not check_ret:
+            test.fail("Unable to find 'Suspended Migrated' event.")
+
+    def cancel_migration_performphase():
```
Comment (dzhengfy) [inline libvirt/tests/src/migration/migrate_options_shared.py:1006]: This function is not required. Refer to below usage:
`action_during_mig = '[{"func": "virsh.domjobabort", "before_pause": "no", "func_param": "'%s' % params.get('migrate_main_vm')"}, {"func": "check_events", "before_pause": "no", "func_param": xxx}]'`

File: libvirt/tests/src/migration/migrate_options_shared.py (hunk 4 above)
Comment (cliping) [reply libvirt/tests/src/migration/migrate_options_shared.py]: Updated. And move this case to a new file. Please help to review, thanks.

File: libvirt/tests/src/migration/migrate_options_shared.py (hunk 6 above)
Comment (cliping) [reply libvirt/tests/src/migration/migrate_options_shared.py]: Updated.

File: libvirt/tests/cfg/migration/migrate_options_shared.cfg (hunk 5 above)
Comment (cliping) [reply libvirt/tests/cfg/migration/migrate_options_shared.cfg]: expectedEventSrc and expectedEventTarget in migrate_options_shared.py: line 1516 to line 1524.

**Comment by cliping** [comment]
Test result:

#avocado run --vt-type libvirt --vt-machine-type q35 virsh.migrate_event.negative_test.cancel_in_performphase.after_src_paused --vt-connect-uri qemu:///system
WARNING:root:No python imaging library installed. Screendump and Windows guest BSOD detection are disabled. In order to enable it, please install python-imaging or the equivalent for your distro.
WARNING:root:No python imaging library installed. PPM image conversion to JPEG disabled. In order to enable it, please install python-imaging or the equivalent for your distro.
JOB ID     : 3fc1f1f0f0f8dd769e91048ed15200e445ffccf8
JOB LOG    : /root/avocado/job-results/job-2021-08-06T06.55-3fc1f1f/job.log
 (1/1) type_specific.io-github-autotest-libvirt.virsh.migrate_event.negative_test.cancel_in_performphase.after_src_paused.p2p_live.without_postcopy: PASS (137.84 s)
RESULTS    : PASS 1 | ERROR 0 | FAIL 0 | SKIP 0 | WARN 0 | INTERRUPT 0 | CANCEL 0
JOB TIME   : 139.74 s

File: libvirt/tests/cfg/migration/migrate_event.cfg [hunk 7]
```diff
@@ -0,0 +1,38 @@
+- virsh.migrate_event:
```
Comment (chunfuwen) [inline libvirt/tests/cfg/migration/migrate_event.cfg:1]: I assume cfg under virsh_cmd begin with virsh, so better find one more proper name matching migration folder

File: provider/migration/migration_base.py [hunk 8]
```diff
@@ -125,3 +127,68 @@ def cleanup_conn_obj(obj_list, test):
         if one_conn:
             logging.debug("Clean up one connection object")
             del one_conn
+
+
+def monitor_event(params):
```
Comment (chunfuwen) [inline provider/migration/migration_base.py:132]: virttest/migration.py, in this file, it sounds like there are some methods to monitor event, can we reuse it?

File: libvirt/tests/cfg/migration/migrate_event.cfg (hunk 7 above)
Comment (cliping) [reply libvirt/tests/cfg/migration/migrate_event.cfg:1]: The first line of other cfg files in migration folder start with virsh.XXX. So I think there's no need to match migration folder.  For example, in cpu folder, the first line of cfg files also don't match cpu folder.

File: provider/migration/migration_base.py (hunk 8 above)
Comment (cliping) [reply provider/migration/migration_base.py:132]: This case need to monitor source and target event. And need to check multiple events. In virttest/migration.py, it's only monitor source event.

File: provider/migration/migration_base.py [hunk 9]
```diff
@@ -125,3 +127,68 @@ def cleanup_conn_obj(obj_list, test):
...
+        virsh_session = virsh.VirshSession(virsh_exec=virsh.VIRSH_EXEC,
+                                           auto_close=True)
+        virsh_session.sendline(cmd)
+
+    if expected_event_target:
+        logging.debug("Running virsh command on target: %s", cmd)
+        remote_session = remote.remote_login("ssh", server_ip, "22",
+                                             server_user, server_pwd,
+                                             r'[$#%]')
+        remote_session.sendline("virsh " + cmd)
```
Comment (Yingshun) [inline provider/migration/migration_base.py:159]: I recommend that you change remote_session to 'remote_virsh_session' and create it via VirshSession() as you did for localhost. 
```
>>> virsh_dargs = {'remote_ip': 'xxx', 'remote_user': 'root','remote_pwd': 'xxx', 'unprivileged_user': None, 'virsh_exec':virsh.VIRSH_EXEC, 'auto_close':'True', 'uri':'qemu+ssh://xxx/system'}
>>> session = virsh.VirshSession(**virsh_dargs)
>>> session.sendline("event --all --loop")
>>> session.get_stripped_output()

```

File: provider/migration/migration_base.py [hunk 10]
```diff
@@ -125,3 +127,68 @@ def cleanup_conn_obj(obj_list, test):
...
+    """
+    logging.debug("Actual output is %s", output)
+    for item in expected_value_list:
+        if not re.findall(item, output):
+            test.fail("Unalbe to find {}".format(item))
+
+
+def check_event_output(params, virsh_session, remote_session, test):
+    """
+    Check event on source/target host
```
Comment (Yingshun) [inline provider/migration/migration_base.py:180]: Please add one blank line Before L181.

File: provider/migration/migration_base.py [hunk 11]
```diff
@@ -125,3 +127,68 @@ def cleanup_conn_obj(obj_list, test):
...
+    :param test: test object
+    :raise: test.fail if unable to find item(s)
+    """
+    logging.debug("Actual output is %s", output)
+    for item in expected_value_list:
+        if not re.findall(item, output):
+            test.fail("Unalbe to find {}".format(item))
+
+
+def check_event_output(params, virsh_session, remote_session, test):
```
Comment (Yingshun) [inline provider/migration/migration_base.py:178]: The default values of virsh_session/ remote_session are None in monitor_event(). So how about setting default value for them in this function and check if they are None before getting the outputs?

File: libvirt/tests/src/migration/migrate_event.py [hunk 12]
```diff
@@ -0,0 +1,97 @@
...
+
+from virttest.libvirt_xml import vm_xml
+from virttest.utils_test import libvirt
+
+from provider.migration import migration_base
+
+
+def run(test, params, env):
+    """
+    Run the test
```
Comment (Yingshun) [inline libvirt/tests/src/migration/migrate_event.py:14]: Could you please update it to more specific one? Thanks!

File: provider/migration/migration_base.py [hunk 13]
```diff
@@ -125,3 +127,68 @@ def cleanup_conn_obj(obj_list, test):
...
+    if expected_event_target:
+        logging.debug("Running virsh command on target: %s", cmd)
+        remote_session = remote.remote_login("ssh", server_ip, "22",
+                                             server_user, server_pwd,
+                                             r'[$#%]')
+        remote_session.sendline("virsh " + cmd)
+    return virsh_session, remote_session
+
+
+def check_output(output, expected_value_list, test):
```
Comment (Yingshun) [inline provider/migration/migration_base.py:163]: Is this function only used for migration test?

File: provider/migration/migration_base.py (hunk 9 above)
Comment (cliping) [reply provider/migration/migration_base.py]: Update.

File: provider/migration/migration_base.py (hunk 11 above)
Comment (cliping) [reply provider/migration/migration_base.py]: Update.

File: libvirt/tests/src/migration/migrate_event.py (hunk 12 above)
Comment (cliping) [reply libvirt/tests/src/migration/migrate_event.py]: Update.

File: provider/migration/migration_base.py (hunk 10 above)
Comment (cliping) [reply provider/migration/migration_base.py:180]: Update.

File: provider/migration/migration_base.py (hunk 13 above)
Comment (Yingshun) [reply provider/migration/migration_base.py:162]: Ok, I don't think it's a blocker for this PR, I'll approve it first as others are good to me now.

**Review by Yingshun — APPROVED** [review]

**Review by dzhengfy — APPROVED** [review]
LGTM

**Review by chunfuwen — APPROVED** [review]
LGTM
